cfgrammar/src/lib/markmap.rs

```diff
@@ -0,0 +1,780 @@
+use quote::{quote, ToTokens};
+use std::borrow::Borrow;
+
+/// MarkMap is a key value data structure that uses an API similar to that of
+/// `std::collections::HashMap` and `std::collections::BTreeMap`.
+///
+/// The current implementation is based on a sorted `Vec` is not optimized for
+/// storing large number of items.
+///
+/// On top of the familiar `std::collections` API it has a few additions:
+///
+/// * Marking a key with a condition.
+/// * Marking a key with a merge behavior.
+/// * A merge operator.
+///
+/// The current *conditions* are [`Used`](MarkMap::mark_used) and [`Required`](MarkMap::mark_required).
+///
+/// The available merge behaviors are [`Theirs`](MergeBehavior::Theirs), [`Ours`](MergeBehavior::Ours),
+/// and [`MutuallyExclusive`](MergeBehavior::MutuallyExclusive).
+///
+/// Merge behaviors configure how the merge operator handles cases where both `MarkMaps` being merged
+/// contain a particular key.
+#[derive(Debug, PartialEq, Eq)]
+pub struct MarkMap<K, V> {
+    contents: Vec<(K, u16, Option<V>)>,
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy)]
+pub enum MergeBehavior {
+    /// The value in `self` takes precedence.
+    Theirs = 1 << 0,
+    /// The value in `other` takes precedence.
+    Ours = 1 << 1,
+    /// It is an error if the key is present in both `MarkMaps`.
+    /// This is the default if no merge behavior is set.
+    MutuallyExclusive = 1 << 2,
+}
+
+#[derive(Debug)]
+pub enum MergeError<K, V> {
+    // Contains the key which was present in both, and the value which was present in `Theirs`.
+    Exclusivity(K, V),
+}
+
+pub enum Entry<'a, K, V> {
+    Occupied(OccupiedEntry<'a, K, V>),
+    Vacant(VacantEntry<'a, K, V>),
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy)]
+enum Mark {
+    // There are some other interesting marks that could be added based on row polymorphic records.
+    // Marks such as `Prohibited` could be used to prove disjointedness.
+    Used,
+    Required,
+    MergeBehavior(MergeBehavior),
+}
+
+impl Mark {
+    fn repr(self) -> u16 {
+        match self {
+            Self::Used => 1 << 0,
+            Self::Required => 1 << 1,
+            Self::MergeBehavior(mb) => (1 << 2) | ((mb as u16) << 8),
```

Comment:
It is probably worth noting that `MergeBehavior(_)`  and it's inner element are both using more bits than necessary, I felt it kept the logic uniform to just give each behavior a 1 bit, and since these are opaque, and not pressed for space, it seemed like we could change it at any time...

I don't know if it is worth a comment, if we were pressed for space we could:
* Drop the `1 << 2` entirely for `MergeBehavior(_)`.
* Just make `MutuallyExclusive` always a zero bit.